text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,149 @@
+ * https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/highlight.rs#L725-L753 (for generating links)
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+[Github Code navigation feature](https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github): Relying on external (and private) tools doesn't seem like a good idea. If they decide to shut it down, we can't do anything about it. In addition to that, a lot of projects are not on github, so it would exclude them. It also requires an internet connection.
+
+rust-analyzer: This is a public opensource project. However, the setup isn't super easy (but it's a minor issue) and when using it on the rust compiler source code, it's quite resource-heavy. Also, if you want to look at a crate source code on <docs.rs>, you'd need to clone it locally to be able to use rust-analyzer on it.
+
+rust-analyzer in the browser: Another possibility would be to run it with wasm. However, it brings a few issues: wasm support in browsers is quite recent, it requires JS to work (obviously, but it's not a big problem), it would be quite heavy to run and it would complexify rustdoc build quite a lot.
+
+# Prior art
+[prior-art]: #prior-art
+
+As mentionned above, it is actually quite common for documentation tools to have a source viewer. Implementations vary a bit between them:
+
+| tool/language | source code viewer | jump to definition | go back to documentation |
+| ------------- | ------------------ | ------------------ | ------------------------ |
+| [doxygen](https://eigen.tuxfamily.org/dox/Matrix_8h_source.html) | yes | no | yes |
+| [Elixir](https://elixir.bootlin.com) (used for linux kernel) | yes | yes | no |
+| [Go](https://cs.opensource.google/go/go/+/go1.16.5:src/database/sql/sql.go;l=44) | yes | yes | no |
+| [hexdocs](https://preview.hex.pm/preview/k6/0.0.1/show/lib/k6/template/web_socket.ex) | yes | no | no |
+| python (sphynx/pydoc) | no (links to github) | no | no |
+| [ruby](https://ruby-doc.org/stdlib-3.1.2/libdoc/delegate/rdoc/Delegator.html#method-i-__raise__) | yes (directly on the page) | yes/no (it is sometimes in examples) | no |
+| [perl](https://metacpan.org/dist/DBD-SQLite/source/lib/DBD/SQLite.pm) | yes | yes | yes |
+| [haskell](https://devdocs.io/haskell~9/libraries/binary-0.8.9.0/data-binary-builder#v:append) | yes | yes | no |
+| [opal](https://opalrb.com/docs/api/v1.4.1/stdlib/Buffer) | yes (directly on the page) | no | yes |
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Is it the best way to make the links to go to an item documentation like this? Maybe something adding a small icon to click right besides it should be better?
+
+Should we add links to documentation to all items (impl blocks, methods, trait/impl associated items) or not?
```

Comment:
I think all items in source code that have documentation should have a way to go from the source view to the documentation.

Reply:
I think so too. I'll move it.